crates/resolver-tests/src/lib.rs

```diff
@@ -280,147 +445,206 @@ impl SatResolve {
+
+        solver.assume(&assumption);
+
+        solver
             .solve()
             .expect("docs say it can't error in default config")
     }
-    pub fn sat_is_valid_solution(&self, pids: &[PackageId]) -> bool {
-        let mut s = self.0.borrow_mut();
-        for p in pids {
-            if p.name().as_str() != "root" && !s.var_for_is_packages_used.contains_key(p) {
+
+    pub fn sat_is_valid_solution(&mut self, pkgs: &[(PackageId, Vec<InternedString>)]) -> bool {
+        let contains_pkg = |pkg| pkgs.iter().any(|(p, _)| p == pkg);
+        let contains_pkg_feature =
+            |pkg, f| pkgs.iter().any(|(p, flist)| p == pkg && flist.contains(f));
+
+        for (p, _) in pkgs {
+            if p.name() != "root" && !self.var_for_is_packages_used.contains_key(p) {
                 return false;
             }
         }
-        let assumption: Vec<_> = s
-            .var_for_is_packages_used
-            .iter()
-            .map(|(p, v)| v.lit(pids.contains(p)))
-            .collect();
 
-        s.solver.assume(&assumption);
+        // root vars from previous runs are deactivated
+        let assumption = (self.old_root_vars.iter().map(|v| v.negative()))
+            .chain(
+                self.var_for_is_packages_used
+                    .iter()
+                    .map(|(p, v)| v.lit(contains_pkg(p))),
+            )
+            .chain(
+                self.var_for_is_packages_features_used
+                    .iter()
+                    .flat_map(|(p, fmap)| {
+                        fmap.iter()
+                            .map(move |(f, v)| v.lit(contains_pkg_feature(p, f)))
+                    }),
+            )
+            .collect::<Vec<_>>();
 
-        s.solver
+        self.solver.assume(&assumption);
+
+        self.solver
             .solve()
             .expect("docs say it can't error in default config")
     }
-    fn use_packages(&self) -> Option<String> {
-        self.0.borrow().solver.model().map(|lits| {
+
+    fn used_packages(&self) -> Option<String> {
+        self.solver.model().map(|lits| {
             let lits: HashSet<_> = lits
                 .iter()
                 .filter(|l| l.is_positive())
                 .map(|l| l.var())
                 .collect();
-            let mut out = String::new();
-            out.push_str("used:\n");
-            for (p, v) in self.0.borrow().var_for_is_packages_used.iter() {
+
+            let mut used_packages = BTreeMap::<PackageId, BTreeSet<InternedString>>::new();
+            for (&p, v) in self.var_for_is_packages_used.iter() {
                 if lits.contains(v) {
-                    writeln!(&mut out, "    {}", p).unwrap();
+                    used_packages.entry(p).or_default();
                 }
             }
+            for (&p, map) in &self.var_for_is_packages_features_used {
+                for (&f, v) in map {
+                    if lits.contains(v) {
+                        used_packages.entry(p).or_default().insert(f);
```

Comment:
Super nit: If we ever use the `or_default` something's gone very wrong. Because we've activated a feature for package that was not activated. And this will be hard to debug because the output will claim that it was activated. I think I'd rather see a panic here. `.entry(p).or_default()` -> `.get_mut(&p).expect("The feature is activated without the package being activated")`